[HttpKernel] Enhance exception if possible

lyrixx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | restore an old feature
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

I don't know if it's the best place to do that.
I think in the kernel, it would be better, WDYT?


Before 

![image](https://github.com/symfony/symfony/assets/408368/e2a452aa-b669-4598-8b23-5366c5b80a74)


After : 

![image](https://github.com/symfony/symfony/assets/408368/0e30c50d-5c94-43db-a9e5-19be6ccd5f2b)

lyrixx committed:
[HttpKernel] Enhance exception if possible

GromNaN commented on src/Symfony/Component/ErrorHandler/ErrorHandler.php:
```diff
@@ -664,9 +664,11 @@ private function renderException(\Throwable $exception): void
     /**
      * Override this method if you want to define more error enhancers.
      *
+     * @internal
```
Are there other places where an `@internal` function is used in another package?

yceruto commented on src/Symfony/Component/HttpKernel/Controller/ErrorController.php:
```diff
@@ -38,6 +39,13 @@ public function __construct(HttpKernelInterface $kernel, string|object|array|nul
+                break;
+            }
+        }
+
```
Before this PR, the error enhancement depends on the `framework.handle_all_throwables` config, if `true` the `Error` instance is rendered through the error-renderer system directly (skipping the error enhancement), otherwise the `ErrorHandler::handleException()` will address it (the error enhancement happens here) and then it will go through the error-renderer system. That's why it's happening. If you change the config to `false` you will see the enhanced error.

Now I'm wondering if it's better to add a new static method inside the ErrorHandler to do this job, thus avoiding duplicated logic (I don't remember the reason but it's currently skipping instances of `OutOfMemoryError`).

Maybe something like this:
```php
// ErrorHandler class
public static enhanceError($error: \Throwable): \Throwable 
{
  if ($exception instanceof OutOfMemoryError) {
     return $error;
  }

  foreach (self::getErrorEnhancers() as $errorEnhancer) {
    if ($error = $errorEnhancer->enhance($exception)) {
      return $error;
    }
  }

  return $error;
}
```
and reuse it in both sides.

stof replied:
in case of an OutOfMemoryError, enhancing it is likely to break due to needing more memory to compute the enhancements. That's why it is skipped.

lyrixx replied:
I was thinking the same things. I'll move the logic

lyrixx replied:
Why is framework.handle_all_throwables false by default?

lyrixx commented:
I really wonder if the controller is the best place to do that 

* If one use a ccustom controller, it won't work
* I should also add support for the console components, right?

lyrixx commented:
I updated the PR

* Introduced `ErrorHandler::enchanceError()`
* Move the call from the ErrorController to the ErrorListener (this is more global like this)

lyrixx commented:
```
There were 15 errors:

1) Symfony\Component\HttpKernel\Tests\EventListener\ErrorListenerTest::testHandleWithoutLogger with data set #0 (Symfony\Component\HttpKernel\Event\ExceptionEvent Object (...), Symfony\Component\HttpKernel\Event\ExceptionEvent Object (...))
Error: Call to undefined method Symfony\Component\ErrorHandler\ErrorHandler::enhanceError()
```

tests are KO because the new code does not exist yet in 7.0?

yceruto commented:
> tests are KO because the new code does not exist yet in 7.0?

I think after synchronizing with 7.0 it should be fixed

nicolas-grekas commented on src/Symfony/Component/ErrorHandler/ErrorHandler.php:
```diff
@@ -661,12 +654,27 @@ private function renderException(\Throwable $exception): void
      * @return ErrorEnhancerInterface[]
      */
-    protected function getErrorEnhancers(): iterable
+    protected static function getErrorEnhancers(): iterable
```
BC break ahead, isn't it?
Can't we get the ErrorHandler instance instead in ErrorListener? It must be somewhere for us to use.

stof replied:
this is indeed a BC break.

lyrixx replied:
Fixed

yceruto commented on src/Symfony/Component/ErrorHandler/ErrorHandler.php:
```diff
@@ -516,14 +516,7 @@ public function handleException(\Throwable $exception): void
-                }
-            }
-        }
+        $exception = self::enhanceError($exception);
```
this should be `$this` instead of `self`, right?

so if a sub-class is overriding the `enhanceError()` method, it will be called for sure.

lyrixx replied:
of course, I forgot to revert this change! Thanks (fixed)

fabpot commented:
Thank you @lyrixx.
